File: pkg/api/activitylogger.go
/*
Copyright Avast Software. All Rights Reserved.

SPDX-License-Identifier: Apache-2.0
*/

package api

import (
	"time"

	"github.com/google/uuid"
)

// Params represents additional parameters which may be required for wallet applications in the future.
// As such, this is currently a placeholder.
type Params map[string]interface{}

// Data represents specific details of an activity.
type Data struct {
	Client    string `json:"client,omitempty"`
	Operation string `json:"operation,omitempty"`
	Status    string `json:"status,omitempty"`
	Params    Params `json:"params,omitempty"`
}

// An Activity represents a single activity.
type Activity struct {
	ID   uuid.UUID `json:"id,omitempty"`
	Type string    `json:"type,omitempty"`
	Time time.Time `json:"time"`
	Data Data      `json:"data"`
}

const (
	// LogTypeCredentialActivity is the string used in an activity to indicate that it is related to a
	// credential operation.
	LogTypeCredentialActivity = "credential-activity" //nolint:gosec // false positive
	// ActivityLogStatusSuccess is the string used in log entries indicating a successful operation.
	ActivityLogStatusSuccess = "success"
)

// An ActivityLogger logs activities.
type ActivityLogger interface {
	// Log logs a single activity.
	Log(credential *Activity) error
}

File: demo/app/lib/models/activity_logger.dart
/*
Copyright Gen Digital Inc. All Rights Reserved.

SPDX-License-Identifier: Apache-2.0
*/

class ActivityLogger {
  final String status;
  final String operation;
  final String date;
  final String issuedBy;
  final String activityType;

  ActivityLogger(
      {required this.status,
      required this.operation,
      required this.date,
      required this.issuedBy,
      required this.activityType});

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = <String, dynamic>{};
    data['Status'] = status;
    data['Operation'] = operation;
    data['Date'] = date;
    data['Issued By'] = issuedBy;
    data['Activity Type'] = activityType;
    return data;
  }

  factory ActivityLogger.fromJson(Map<String, dynamic> json) {
    return ActivityLogger(
      status: json['Status'],
      operation: json['Operation'],
      date: json['Date'],
      issuedBy: json['Issued By'],
      activityType: json['Activity Type'],
    );
  }
}
